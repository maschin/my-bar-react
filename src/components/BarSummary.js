import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const BarSummary = ({ bottlesCount }) => {
    const bottlesWord = bottlesCount === 1 ? 'bottle' : 'bottles';

    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    You have total <span>{bottlesCount}</span> {bottlesWord} in your bar.
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Drink</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks,
        bottlesCount: state.drinks.length
    };
};

export default connect(mapStateToProps)(BarSummary);