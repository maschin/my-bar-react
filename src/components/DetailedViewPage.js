import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
    openDrink,
    emptyDrink
} from '../actions/drinks';

export class DetailedView extends React.Component {
    onOpen = () => {
        this.props.openDrink(this.props.drink);
    }
    onMakeEmpty = () => {
        this.props.emptyDrink(this.props.drink);
    }
    render(){
        return(
            <div>
                {
                    this.props.drink ? (
                        <div>
                            <div className="page-header">
                                <div className="content-container">
                                    <div className="page-header__actions">
                                        <Link className="button" to={`/edit/${this.props.drink.id}`}>Edit</Link>
                                        { !this.props.drink.isOpen ? (
                                            <button id="openDrink" className="button" onClick={this.onOpen}>Open</button>
                                        ) : (

                                            !this.props.drink.isEmpty && <button id="emptyDrink" className="button" onClick={this.onMakeEmpty}>Empty</button>

                                        )}
                                    </div>
                                </div>
                            </div>
                            <div className="content-container">
                                <h1>{ this.props.drink.title }</h1>
                                <ul>
                                    <li>{ this.props.drink.category }</li>
                                    <li>{ this.props.drink.vol } L</li>
                                    <li>{ this.props.drink.alc } %</li>
                                    <li>{ this.props.drink.price }</li>
                                </ul>
                            </div>
                        </div>
                    ) : (
                        <h1 className="page-header__title">Record with <span>id:{this.props.match.params.id}</span> not found!</h1>
                )}
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        drink: state.drinks.find((drink) => (drink.id === props.match.params.id))
    }
}

const mapDispatchToProps = (dispatch) => ({
        openDrink: (drink) => dispatch(openDrink(drink)),
        emptyDrink: (drink) => dispatch(emptyDrink(drink))
})

export default connect(mapStateToProps, mapDispatchToProps)(DetailedView);