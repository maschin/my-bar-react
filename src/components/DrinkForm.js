import React from 'react';
import moment from 'moment';
import drinks from '../tests/fixtures/drinks';
import {SingleDatePicker} from 'react-dates';

export class DrinkForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.drink ? props.drink.title : '',
            category: props.drink ? props.drink.category : '',
            vol: props.drink ? props.drink.vol : '',
            alc: props.drink ? props.drink.alc : '',
            price: props.drink ? (props.drink.price / 100).toString() : '',
            aged: props.drink ? props.drink.aged : '',
            description: props.drink ? props.drink.description : '',
            isEmpty: props.drink ? props.drink.isEmpty : false,
            isOpen: props.drink ? props.drink.isEmpty : false,
            boughtAt: props.drink ? moment(props.drink.boughtAt) : null,
            bottledAt: props.drink ? moment(props.drink.bottledAt) : null,
            bestBefore: props.drink ? moment(props.drink.bestBefore) : null,
            createdAt: props.drink ? moment(props.drink.createdAt) : moment(),
            updatedAt: props.drink ? moment(props.drink.updatedAt) : null,
            errors: []
        };
    };
    onFieldChange = (e, field) => {
        const value = e.target.value;
        switch(field){
            case 'price':
                this.onPriceChange(e);
                break;
            default:
                this.setState({[field]: value});
                break;
        }
    };
    onPriceChange = (e) => {
        const price = e.target.value;
        if (!price || price.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({price}));
        }
    };
    onCalendarDateChange = (date, field) => {
        this.setState({[field]: date});
    };
    onCalendarFocusChange = (focused, field) => {
        field = `calendar${field}Focused`;
        this.setState(() => ({[field]: focused}));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (this.validate()) {
            this.props.onSubmit({
                title: this.state.title,
                category: this.state.category,
                vol: this.state.vol,
                alc: this.state.alc,
                price: this.state.price ? parseFloat(this.state.price, 10) * 100 : '',
                aged: this.state.aged,
                description: this.state.description,
                isEmpty: this.state.isEmpty,
                isOpen: this.state.isOpen,
                boughtAt: this.state.boughtAt.valueOf(),
                bottledAt: this.state.bottledAt.valueOf(),
                bestBefore: this.state.bestBefore.valueOf(),
                createdAt: this.state.createdAt.valueOf(),
                updatedAt: moment().valueOf()
            });
            console.log(this.state.price);
        }
    };
    validate = () => {
        let errors = [];

        if(this.props.options.validate) {

            if (!this.state.title) {
                errors.push('Please provide a title');
            }

            if (!this.state.category) {
                errors.push('Please provide a category');
            }

            if (!this.state.vol) {
                errors.push('Please provide a volume');
            }

            if (!this.state.alc) {
                errors.push('Please provide a alcohol in %');
            }

            if (!this.state.price) {
                errors.push('Please provide a price');
            }
        }

        if (!errors.length)
            return true;
        else
            this.setState({errors})
    };

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {
                    this.state.errors.length !== 0 && this.state.errors.map((error) => (
                        <p className="form__error" key={error}>{error}</p>
                    ))
                }
                <input
                    type="text"
                    className="text-input"
                    placeholder="Title"
                    autoFocus
                    value={this.state.title}
                    onChange={e => this.onFieldChange(e, 'title')}
                />
                <select
                    className="select"
                    value={this.state.category}
                    onChange={e => this.onFieldChange(e, 'category')}
                >
                    <option value=''>-- Please select --</option>
                    {this.props.categories.map((category) => <option key={category}
                                                                     value={category}>
                        {category}
                        </option>)}
                </select>
                <select
                    className="select"
                    value={this.state.vol}
                    onChange={e => this.onFieldChange(e, 'vol')}
                >
                    <option value=''>-- Please select --</option>
                    {this.props.vol.map((v) => <option key={v} value={v}>{v}</option>)}
                </select>
                <input
                    type="text"
                    className="text-input"
                    placeholder="Alcohol"
                    value={this.state.alc}
                    onChange={e => this.onFieldChange(e, 'alc')}
                />
                <input
                    type="text"
                    id="price"
                    className="text-input"
                    placeholder="Price"
                    value={this.state.price}
                    onChange={e => this.onFieldChange(e, 'price')}
                />
                <input
                    type="text"
                    className="text-input"
                    placeholder="Specify aged of drink (optional)"
                    value={this.state.aged}
                    onChange={e => this.onFieldChange(e, 'aged')}
                />
                <textarea
                    className="text-area"
                    placeholder="Description for your drink (optional)"
                    value={this.state.description}
                    onChange={e => this.onFieldChange(e, 'description')}
                >
                </textarea>
                <SingleDatePicker
                    date={this.state.boughtAt}
                    required={true}
                    placeholder="Bought date"
                    onDateChange={date => this.onCalendarDateChange(date, 'boughtAt')}
                    focused={this.state.calendarBoughtAtFocused}
                    onFocusChange={({focused}) => this.onCalendarFocusChange(focused, 'BoughtAt')}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <SingleDatePicker
                    date={this.state.bottledAt}
                    required={true}
                    placeholder="Bottled date"
                    onDateChange={date => this.onCalendarDateChange(date, 'bottledAt')}
                    focused={this.state.calendarBottledAtFocused}
                    onFocusChange={({focused}) => this.onCalendarFocusChange(focused, 'BottledAt')}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <SingleDatePicker
                    date={this.state.bestBefore}
                    placeholder="Best before"
                    required={true}
                    onDateChange={date => this.onCalendarDateChange(date, 'bestBefore')}
                    focused={this.state.calendarBestBeforeFocused}
                    onFocusChange={({focused}) => this.onCalendarFocusChange(focused, 'BestBefore')}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <div>
                    <button className="button">{this.props.actionTitle}</button>
                </div>
            </form>
        );
    }
}

export default DrinkForm;

DrinkForm.defaultProps = {
    options: { validate: true },
    categories: ['Beer', 'Whisky'],
    vol: [0.33, 0.5, 0.7, 1]
};