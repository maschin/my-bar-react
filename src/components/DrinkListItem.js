import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const DrinkListItem = ({
                                  id, title,
                                  category, vol,
                                  alc, price, aged,
                                  isOpen, isEmpty,
                                  boughtAt, bottledAt, bestBefore,
                              }) => (
    <div>
        <Link className={`list-item ${isOpen ? 'list-item--opened' : ''} ${isEmpty ? 'list-item--empty' : ''}`} to={`/view/${id}`}>
            <div>
                {title}
                <br/>
                {category} - {vol}L
            </div>
            <ul>
                <li>{alc}%</li>
                <li>{numeral(price/100).format('$0,0.00')}</li>
                <li>Age: {aged}</li>
            </ul>
            <ul>
                <li>Bought: {moment(boughtAt).format('MMMM Do, YYYY')}</li>
                <li>Bottled: {moment(bottledAt).format('MMMM Do, YYYY')}</li>
                <li>BBF: {moment(bestBefore).format('MMMM Do, YYYY')}</li>
            </ul>
        </Link>
    </div>
);



export default DrinkListItem;