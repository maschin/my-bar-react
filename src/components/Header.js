import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout, userData }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>My Bar</h1>
                </Link>
                <div>
                    <h4>
                        <a href="https://myaccount.google.com">
                            <img className="header__avatar" src={ userData.photoURL } width="40px" /> &nbsp; Welcome, {userData.name}
                        </a>
                    </h4>
                </div>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

const mapStateToProps = (state) => ({
    userData: state.auth
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);