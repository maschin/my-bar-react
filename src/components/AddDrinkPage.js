import React from 'react';
import { connect } from 'react-redux';

import DrinkForm from '../components/DrinkForm';
import { startAddDrink } from '../actions/drinks';

export class AddDrinkPage extends React.Component {
    onSubmit = (drink) => {
        this.props.startAddDrink(drink);
        this.props.history.push('/');
    };
    render(){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Bottle</h1>
                    </div>
                </div>
                <div className="content-container">
                    <DrinkForm
                        actionTitle='Add Bottle'
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddDrink: (drink) => dispatch(startAddDrink(drink))
});

export default connect(undefined, mapDispatchToProps)(AddDrinkPage);