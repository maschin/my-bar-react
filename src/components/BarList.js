import React from 'react';
import { connect } from 'react-redux';

import DrinkListItem from '../components/DrinkListItem';

export const BarList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-desktop">Drink</div>
            <div className="show-for-desktop">Data</div>
        </div>
        <div className="list-body">
            {
                props.drinks.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>Your bar is totally empty!</span>
                    </div>
                ) : (
                    props.drinks.map((drink) => (
                       <DrinkListItem key={drink.id} {...drink} />
                    ))
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        drinks: state.drinks
    };
};

export default connect(mapStateToProps)(BarList);