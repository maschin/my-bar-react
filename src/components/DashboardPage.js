import React from 'react';
import BarList from '../components/BarList';
import BarSummary from '../components/BarSummary';

export const DashboardPage = () => (
    <div>
        <BarSummary />
        <BarList />
    </div>
);

export default DashboardPage;