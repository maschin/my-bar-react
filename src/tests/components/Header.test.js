import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

let startLogout, wrapper;

beforeEach (() => {
    const userData = {
        name: 'Mike',
        photoURL: 'https://imgsome'
    };
    startLogout = jest.fn();
    wrapper = shallow(<Header startLogout={startLogout} userData={userData} />);
});

test('should render Header correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});


