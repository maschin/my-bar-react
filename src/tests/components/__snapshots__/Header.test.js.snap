// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          My Bar
        </h1>
      </Link>
      <div>
        <h4>
          <a
            href="https://myaccount.google.com"
          >
            <img
              className="header__avatar"
              src="https://imgsome"
              width="40px"
            />
             Â  Welcome, 
            Mike
          </a>
        </h4>
      </div>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
